### ISSUE TITLE
Decimal qty with Increment is with specific values are not adding in cart

### ISSUE THREAD

**Issue by amanuni** [magento/magento2#23038]
### Preconditions (*)

1. Product with Any Decimal Quantity should be added in Cart If Decimal "Qty Uses Decimals" and option is "Yes"
2. Product is not adding in cart with qty decimal, qty increment is enabled and minimum qty assigned.

### Steps to reproduce (*)

1. Create simple product enable "Qty Uses Decimal" and Assign "Minimum Qty" to 0.5 then enable "Enable Qty Increments" and assign 0.1 to "Qty Increments"
2. Here is the case Minimum Qty in Cart 0.5, Qty Increment Is 0.1

### Expected result (*)

1. Product should be added in cart.

### Actual result (*)

1. Product should be added in cart when click on add to cart but its adding and throwing message "You can buy this product only in quantities of 0.1 at a time."

**Comment by amanuni** [comment]
@magento give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @amanuni. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @amanuni, here is your [Magento instance](https://i-23038-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-23038-2-3-develop.instances.magento-community.engineering/admin](https://i-23038-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by amanuni** [comment]
@magento-engcom-team issue reproduced in provided instance. Check below mentioned product i have created and its happening with it.

https://i-23038-2-3-develop.instances.magento-community.engineering/test-decimal-product.html

**Comment by ktpl-rammohan** [comment]
@magento give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @rammohan-krish. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @rammohan-krish, here is your [Magento instance](https://i-23038-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-23038-2-3-develop.instances.magento-community.engineering/admin](https://i-23038-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by amanuni** [comment]
@rammohan-krish thank you

**Comment by ktpl-rammohan** [comment]
@amanuni you're welcome.

**Comment by ktpl-rammohan** [comment]
@magento give me 2.2-develop instance

**Comment by magento-engcom-team** [comment]
Hi @rammohan-krish. Thank you for your request. I'm working on Magento 2.2-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @rammohan-krish, here is your [Magento instance](https://i-23038-2-2-develop.instances.magento-community.engineering).
Admin access: [https://i-23038-2-2-develop.instances.magento-community.engineering/admin](https://i-23038-2-2-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @rammohan-krish 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-17105`, `MC-17106` were created

**Issue Available**: @rammohan-krish, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by amanuni** [comment]
@rammohan-krish @magento-engcom-team i know why this issue is appearing and what is the problem i have added a solution for it.

Here is problem why issue is appearing.
in file `Magento_Root//lib/web/mage/validation.js:1606`

`isQtyIncrementsValid = typeof params.qtyIncrements === 'undefined' ||
                        qty % $.mage.parseNumber(params.qtyIncrements) === 0;`

This line is causing issue actually this happening wrong decimal calculations of binary numbers. Check the answer in this thread.

https://stackoverflow.com/questions/3966484/why-does-modulus-operator-return-fractional-number-in-javascript

according to my provided value dry run will be like that.

`isQtyIncrementsValid = typeof params.qtyIncrements === 'undefined' ||
                        0.5 % 0.1 === 0;`
The general calculation result is 0 but.
in computer its returning decimal number 0.09999999999999998 so its throwing the error.

Here is the screen shot of calculation seen  screen shot console.
![image](https://user-images.githubusercontent.com/2352740/58639029-bd984100-8313-11e9-9273-caeb9aadd628.png)

So its returning a value and causing error.

Here is the solution i have implemented to resolve this issue.

```
'validate-item-quantity': [
            function (value, element, params) {
                var validator = this,
                    result = false,
                    // obtain values for validation
                    qty = $.mage.parseNumber(value),
                    isMinAllowedValid = typeof params.minAllowed === 'undefined' ||
                        qty >= $.mage.parseNumber(params.minAllowed),
                    isMaxAllowedValid = typeof params.maxAllowed === 'undefined' ||
                        qty <= $.mage.parseNumber(params.maxAllowed),
//Edit started
                    isQtyIncrementsValid;
if(!/^-?\d+$/.test(qty) || !/^-?\d+$/.test($.mage.parseNumber(params.qtyIncrements))){
					isQtyIncrementsValid = typeof params.qtyIncrements === 'undefined' || floatSafeModulus(qty,$.mage.parseNumber(params.qtyIncrements))===0;
				}else{
					isQtyIncrementsValid = typeof params.qtyIncrements === 'undefined' ||
                        qty % $.mage.parseNumber(params.qtyIncrements) === 0;
				}
//Edit Ended
                result = qty > 0;

                if (result === false) {
                    validator.itemQtyErrorMessage = $.mage.__('Please enter a quantity greater than 0.');//eslint-disable-line max-len

                    return result;
                }

                result = isMinAllowedValid;

                if (result === false) {
                    validator.itemQtyErrorMessage = $.mage.__('The fewest you may purchase is %1.').replace('%1', params.minAllowed);//eslint-disable-line max-len

                    return result;
                }

                result = isMaxAllowedValid;

                if (result === false) {
                    validator.itemQtyErrorMessage = $.mage.__('The maximum you may purchase is %1.').replace('%1', params.maxAllowed);//eslint-disable-line max-len

                    return result;
                }

                result = isQtyIncrementsValid;

                if (result === false) {
                    validator.itemQtyErrorMessage = $.mage.__('You can buy this product only in quantities of %1 at a time.').replace('%1', params.qtyIncrements);//eslint-disable-line max-len

                    return result;
                }

                return result;
            }, function () {
                return this.itemQtyErrorMessage;
            }
        ]
```
//Function added in file

```
function floatSafeModulus(val, step) {
		var valDecCount = (val.toString().split('.')[1] || '').length,
		stepDecCount = (step.toString().split('.')[1] || '').length,
		decCount = valDecCount > stepDecCount ? valDecCount : stepDecCount,
		valInt = parseInt(val.toFixed(decCount).replace('.', '')),
		stepInt = parseInt(step.toFixed(decCount).replace('.', ''));
		return (valInt % stepInt) / Math.pow(10, decCount);
}

```

Also i have added edited validation.js file hope you guys may get help with its if working perfectly for me
[validation.zip](https://github.com/magento/magento2/files/3237404/validation.zip)

**Comment by amanuni** [comment]
@magento-engcom-team any update on it.

**Comment by sertlab** [comment]
#mmuk2019

**Comment by magento-engcom-team** [comment]
@sertlab thank you for joining. Please accept team invitation [here](https://github.com/orgs/magento/invitation) and self-assign the issue.

**Comment by sertlab** [comment]
#mmuk2019 assign to me

**Comment by amanuni** [comment]
Thank you guys for working on this issue.

**Comment by magento-engcom-team** [comment]
Hi @amanuni. Thank you for your report. 
The issue has been fixed in magento/magento2#23574 by @krishprakash in 2.2-develop branch
Related commit(s):
 - [0b5ee99cf568d29829b2e1b638283f7807c2966d](https://github.com/magento/magento2/commit/0b5ee99cf568d29829b2e1b638283f7807c2966d)
 - [138b4cf31c6d499d2817c53c663710ce894c058a](https://github.com/magento/magento2/commit/138b4cf31c6d499d2817c53c663710ce894c058a)

The fix will be available with the upcoming 2.2.10 release.

**Comment by amanuni** [comment]
Thank you very much @sertlab its really appreciated.
